src/components/SignUpForm.js

```diff
@@ -0,0 +1,167 @@
+import React, { Component } from 'react'
+import zxcvbn from 'zxcvbn'
+import { validate } from '../helpers/validations'
+import { Button, Form, Label, Progress } from 'semantic-ui-react'
+import _ from 'lodash'
+
+const PasswordStrengthBar = ({ score }) =>
+  <Progress
+    attached="bottom"
+    title={`Password strength: ${[
+      'Very weak',
+      'Weak',
+      'Okay',
+      'Strong',
+      'Very strong'
+    ][score]}`}
+    style={{ marginTop: -3 }}
+    percent={score * 25}
+    color={['red', 'orange', 'yellow', 'olive', 'green'][score]}
+  />
+
+export class SignUpForm extends Component {
+  constructor(props) {
+    super(props)
+    this.fields = {}
+    this.state = {
+      passwordScore: 0,
+      errors: {}
+    }
+  }
+  componentDidMount() {
+    this._isMounted = true
+  }
+  componentWillUnmount() {
+    this._isMounted = false
+  }
+  fieldValues() {
+    return _.mapValues(this.fields, (f) => f && f.value)
+  }
+  onSubmit(e) {
+    e.preventDefault()
+    let fieldValues = this.fieldValues(),
+        validationErrors = validate(this.props.validations, fieldValues)
+    return validationErrors.then((errors) => {
+      this.setState({
+        errors: errors
+      })
+
+      if (Object.keys(errors).length === 0) {
+        return this.props
+          .onSubmit(fieldValues)
+          .then(this.props.onSuccess, this.props.onError)
+      } else {
+        return this.props.onError()
+      }
+    })
+  }
+  validateField(fieldName) {
+    return _.debounce(
+      (e) =>
+        validate(
+          this.props.validations,
+          this.fieldValues(),
+          fieldName
+        ).then((errors) => {
+          if (this._isMounted) {
+            this.setState((state) => ({
+              ...state,
```

Comment:
They don't state that it's needed. They only show the possibility to access the previous state to create a new state.

Would be interesting to find out whether or not the `...state` is really needed. Could you please try it and report the result here? :)